restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java

```diff
@@ -881,128 +841,84 @@ private JsonNode searchUsingCassandraFilters(
-            searchRows(
-                keyspace,
-                collection,
-                db,
-                Collections.emptyList(),
-                ImmutableList.of(condition),
-                Collections.emptyList(),
-                path,
-                false,
-                null,
-                paginator);
-        candidatesThisPage = new LinkedHashSet<>();
-        for (Row row : page) {
-          candidatesThisPage.add(row.getString("key"));
-        }
-      } else {
-        LinkedHashSet<String> tempCandidates = new LinkedHashSet<>();
-        for (String candidate : candidatesThisPage) {
-          BuiltCondition candidateCondition = BuiltCondition.of("key", Predicate.EQ, candidate);
-          page =
-              searchRows(
-                  keyspace,
-                  collection,
-                  db,
-                  ImmutableList.of(candidateCondition),
-                  ImmutableList.of(condition),
-                  Collections.emptyList(),
-                  path,
-                  false,
-                  null,
-                  paginator);
-          if (!page.isEmpty()) {
-            tempCandidates.add(candidate);
-          }
-        }
-        candidatesThisPage = tempCandidates;
-      }
+    db.authorizeSelect(keyspace, collection);
+
+    Iterator<FilterCondition> it = inCassandraFilters.iterator();
+    FilterCondition pagingCondition = it.next(); // assume at least one condition
+
+    AbstractBound<?> mainQuery =
+        db.builder()
+            .select()
+            .column("key")
+            .column("leaf")
+            .from(keyspace, collection)
+            .where(buildConditions(pagingCondition, db.treatBooleansAsNumeric()))
+            .allowFiltering()
+            .build()
+            .bind();
+
+    Flowable<RawDocument> docs =
+        db.executeSelect(
+            mainQuery, docsApiConfiguration.getSearchPageSize(), paginator.getCurrentDbPageState());
+
+    while (it.hasNext()) {
+      FilterCondition nestedCondition = it.next();
+
+      // chain the other filters as nested queries
+      docs =
+          docs.concatMap(
+              d -> {
+                BuiltCondition keyCondition = BuiltCondition.of("key", Predicate.EQ, d.id());
+
+                AbstractBound<?> nestedQuery =
+                    db.builder()
+                        .select()
+                        .column("key")
+                        .column("leaf")
+                        .from(keyspace, collection)
+                        .where(keyCondition)
+                        .where(buildConditions(nestedCondition, db.treatBooleansAsNumeric()))
+                        .allowFiltering()
+                        .build()
+                        .bind();
+
+                // If the nested query finds any docs, return the input doc to preserve
+                // the paging order of the main query
+                return db.executeSelect(nestedQuery, 1, null).limit(1).map(nested -> d);
```

Comment:
OK so this look suspicions to me, we would execute additionally MxN queries in this loop for each remaining condition (M) and renaming number of "surviving docs" (N).. so M*N worst case.. M best case.. if we cut half each time it's gonna be M * N/2, still problematic imo..

Can't we fetch all documents containing the given key collection? Not sure if IN would be supported here? `Moreover, the IN relation is only allowed on the last column of the partition key and on the last column of the full primary key.` Isn't the `key` the only partition key?

It requires a bit of the gymnastics here.. but you can create another flow which you would use for the filtering in an async way()..

```
Flowable<RawDocument> surviving = docs.collectList()
.flatMap(docs -> {
   Collection<?> ids =  docs.stream().map(RawDocument::id).collect();
   BuiltCondition keyCondition = BuiltCondition.of("key", Predicate.IN, ids);
   ....
   return db.executeSelect(nestedQuery, docs.size(), null)
});
docs = docs.compose(FlowableTransformers.filterAsync(v -> surviving.any(s -> s.id().equals(v.id())
```

Reply:
Let's test it :)

Reply:
cf. https://github.com/stargate/stargate/pull/937#discussion_r629738175

Reply:
So not possible:

`Select on indexed columns and with IN clause for the PRIMARY KEY are not supported`.

Reply:
Noted. Thanks for testing it out :+1: